File: R/sentence_functions.R
#' Tokenize character vector to a list of lists of sentences
#'
#' IMPORTANT: must be run first in order to create
#'   internal Python object `sentences`
#'   and `sentences_lower`
#'   to be accessed by `get_sentence_filtered_doc_indices()`
#'   (and thereby also by `create_sentence_filtered_df()`
#'   and `explore_sentence_filtered_df()`).
#'
#' @param df R data frame or corporaexplorerobject
#' @param to_lower convert the final list of lists to lower case?
#'
#' @return internal Python object `sentences`
#'   and `sentences_lower`.
#' @export
tokenize_sentences_ru <- function(df, to_lower = TRUE) {
    if (identical(class(df), "corporaexplorerobject")) {
      df <- corporaexplorer:::get_df(df)
    }
    py$Text <- df$Text
    reticulate::py_run_string("sentences = tokenize_sentences_ru(Text)")

    if (to_lower == TRUE) {
        reticulate::py_run_string("sentences_lower = sentences_to_lower(sentences)")
    }
}

#' @export
tokenize_sentences_no <- function(df, to_lower = TRUE) {
    if (identical(class(df), "corporaexplorerobject")) {
      df <- corporaexplorer:::get_df(df)
    }
    py$Text <- df$Text
    reticulate::py_run_string("sentences = tokenize_sentences_no(Text)")

    if (to_lower == TRUE) {
        reticulate::py_run_string("sentences_lower = sentences_to_lower(sentences)")
    }
}

#' @export
tokenize_sentences_en <- function(df, to_lower = TRUE) {
    if (identical(class(df), "corporaexplorerobject")) {
      df <- corporaexplorer:::get_df(df)
    }
    py$Text <- df$Text
    reticulate::py_run_string("sentences = tokenize_sentences_en(Text)")

    if (to_lower == TRUE) {
        reticulate::py_run_string("sentences_lower = sentences_to_lower(sentences)")
    }
}

#' Wrapper for convenient argument passing
#'
#' Important: "silent" argument: `sentences` in the Python process,
#'   created by `tokenize_sentences_wrapper()`
#'
#' @param filter_pattern Chr. around which pattern extract sentences?
#' @param search_pattern Which pattern to search within extracted sentences?
#' @param r_indexing Add 1 to python zero indexing to achieve equivalent
#'   indexing in R?
#' @param window How many sentences to each side of `filter_pattern` to extract?
#'
#' @return R integer vector with (Python zero style) indices for the
#'   documents that have `search_pattern` in `window` sentence proximity
#'     to `filter_pattern`.
#' @export
get_sentence_window_indices <-
  function(filter_pattern,
           window,
           case_sensitive = FALSE,
           indices_included = NULL,
           py_var_name = "filtered_indices") {
    py$filter_pattern <- filter_pattern
    py$window <- as.integer(window)
    py$indices_included <- indices_included

    if (case_sensitive == FALSE) {
      reticulate::py_run_string(sprintf("

%s = sentence_windows_i_for_corpus(sentences_lower, filter_pattern, window, indices_included)
%s = remove_docs_without_hits(%s)

",
py_var_name, py_var_name, py_var_name))

    } else if (case_sensitive == TRUE) {
      reticulate::py_run_string(sprintf("

%s = sentence_windows_i_for_corpus(sentences, filter_pattern, window, indices_included)
%s = remove_docs_without_hits(%s)

",
py_var_name, py_var_name, py_var_name))
    }
  }

#' @export
filter_sentence_window <- function(filter_pattern, case_sensitive = FALSE, py_var_name = "filtered_indices") {
  py$filter_pattern <- filter_pattern

  if (case_sensitive == FALSE) {

  reticulate::py_run_string(sprintf("

%s = filter_index_object(%s, sentences_lower, filter_pattern)
%s = remove_docs_without_hits(%s)

",
py_var_name, py_var_name, py_var_name, py_var_name))

  } else if (case_sensitive == TRUE) {
  reticulate::py_run_string(sprintf("

%s = filter_index_object(%s, sentences, filter_pattern)
%s = remove_docs_without_hits(%s)

",
py_var_name, py_var_name, py_var_name, py_var_name))
  }
}

#' Title
#'
#' @param r_indexing Whether the input is from R index (starting on 1)
get_filtered_doc_indices_from_py <- function(r_indexing = TRUE) {
    indices <- reticulate::py_eval("get_indices(filtered_indices)")

    if (length(indices) > 0) {
        indices <- indices + r_indexing
    }
    return(indices)
}

get_number_of_sentences_per_doc <- function(r_indexing = TRUE) {
    py$chunks_per_doc <- reticulate::py_eval("number_of_extracted_sentences_per_doc(charting_indices)")
    docs <- reticulate::py_eval("get_indices(chunks_per_doc)")
    if (length(docs) > 0) {
        docs <- docs + r_indexing
        n <- reticulate::py_eval("[x[1] for x in chunks_per_doc]")
        df <- data.frame(Term_1 = n, ID = docs)
        return(df)
    }
}

# TODO. Ser stygt ut i
#' Title
get_filtered_sentences_from_one_doc_py <- function() {
    text <- reticulate::py_eval("retrieve_sentences_from_nested_indices_one_doc(doc_indices, sentences)")
    text <- lapply(text, paste, collapse = " ") %>%
      unlist(use.names = FALSE) %>%
        paste(collapse = "\n\n")
}

#' Returns a filtered
#'   (full text) df based on the returned indices
create_sentence_filtered_df <-
    function(df,
             filter_pattern,
             search_pattern,
             window) {

        get_nested_indices(filter_pattern,
                              search_pattern,
                              window)

        df_indices <- get_filtered_doc_indices_from_py()

        new_df <- df[df_indices, ]

        return(new_df)
    }

#' runs corporaexplorer `explore0()` on the df returned by
#'   `create_sentence_filtered_df()`, with two search fields
#'   pre-populated
explore_sentence_filtered_df <-
    function(df,
             filter_pattern,
             search_pattern,
             window) {

        new_df <- create_sentence_filtered_df(df,
                                          filter_pattern,
                                          search_pattern,
                                          window)

        explore0(new_df, arguments_explore = list(search_input = list(
            search_terms = c(filter_pattern, search_pattern)
        )))
    }

File: inst/python/sentence_functions.py
from tqdm import tqdm
import re


def flatten_list(list_of_lists):
    """
    :param list_of_lists: list
        List of unnested lists, e.g. [[1, 2], [3,4]]
    :return: unnested list
    """
    flat_list = [item for sublist in list_of_lists for item in sublist]
    return flat_list


def sentences_to_lower(list_of_lists):
    return [(i,[sentence.lower() for sentence in doc]) for i,doc in list_of_lists]


def remove_docs_without_hits(kwics):
    kwics = [x for x in kwics if len(x[1]) > 0]
    return kwics


def get_indices(x):
    return [_[0] for _ in x]
    

def sentence_window_i(list_of_sentences, word, window):
    kwics = []
    for i, sentence in enumerate(list_of_sentences):
        if re.search(word, sentence):
            first_index = i - window
            if first_index < 0:
                first_index = 0
            second_index = i + 1 + window
            if second_index > len(list_of_sentences):
                second_index = len(list_of_sentences)
            kwics.append(slice(first_index, second_index))
    return kwics


def sentence_windows_i_for_corpus(sentence_object, word, window, indices_included = None):
    if indices_included is not None:
        if type(indices_included) is not list:
            indices_included = [indices_included]
        docs = set(indices_included)
        sentence_object = [x for x in sentence_object if x[0] in docs]
    return [(doc[0], sentence_window_i(doc[1], word, window)) for doc in sentence_object]


def filter_index_object(index_object, sentence_object, search_pattern):
    new_index_object = []
    for index, slices in index_object:
        sublist = []
        for chunk in slices:
            text = " ".join(sentence_object[index][1][chunk])
            if re.search(search_pattern, text):
                sublist.append(chunk)
        new_index_object.append((index, sublist))
    return new_index_object


def retrieve_sentences_from_nested_indices_one_doc(
    index_object_one_doc, sentence_object
):
    assert len(index_object_one_doc) < 2
    setninger_i_dok = []
    
    if len(index_object_one_doc) == 1:
        index_object_one_doc = index_object_one_doc[0]
        doc_index = index_object_one_doc[0]
        
        for chunk in index_object_one_doc[1]:
                setninger_i_dok.append(sentence_object[doc_index][1][chunk])
                
    return setninger_i_dok


def number_of_extracted_sentences_per_doc(index_object):
    chunks_per_doc = [(doc[0], len(doc[1])) for doc in index_object]
    return chunks_per_doc
